cmd/kubelet/app/server.go

```diff
@@ -385,36 +303,78 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
-			if err == nil {
-				// Update s (KubeletServer) with new config from API server
-				s.KubeletConfiguration = *remoteKC
-				// Ensure that /configz is up to date with the new config
-				if cfgzErr != nil {
-					glog.Errorf("was unable to register configz before due to %s, will not be able to set now", cfgzErr)
-				} else {
-					setConfigz(cfgz, &s.KubeletConfiguration)
-				}
-				// Update feature gates from the new config
-				err = utilfeature.DefaultFeatureGate.Set(s.KubeletConfiguration.FeatureGates)
-				if err != nil {
-					return err
-				}
-			} else {
-				glog.Errorf("failed to init dynamic Kubelet configuration sync: %v", err)
+
+	// launch nodeconfig controller if the dynamic config feature gate is enabled
+	dynamicConfigEnabled := utilfeature.DefaultFeatureGate.Enabled(features.DynamicKubeletConfig)
+	if dynamicConfigEnabled {
+		// `dynamicConfigEnabled` indicates the feature gate, which is presently used to gate the whole controller
+		// eventually, this restriction will be reduced to only gating the "dynamic" part of the controller,
+		// which is why `dynamicConfigEnabled` is still included as part of the `useDynamicConfig` predicate.
+		useDynamicConfig := dynamicConfigEnabled && !standaloneMode && !s.RunOnce
+
+		flagCfg := &componentconfigv1alpha1.KubeletConfiguration{}
+		err = api.Scheme.Convert(&s.KubeletConfiguration, flagCfg, nil)
+		if err != nil {
+			return err
+		}
+
+		// the NodeConfigDir must be specified to use the controller
+		if len(s.NodeConfigDir) == 0 {
+			return fmt.Errorf("--node-config-dir must be specified to use the node config controller")
+		}
+		// compute absolute path for NodeConfigDir, if necessary
+		var nodeConfigDir string
+		if s.NodeConfigDir[0] != '/' {
+			nodeConfigDir = filepath.Join(s.RootDirectory, s.NodeConfigDir)
+		} else {
+			nodeConfigDir = s.NodeConfigDir
+		}
+
+		var ncc *nodeconfig.NodeConfigController
+		if useDynamicConfig {
+			client, err := getKubeClient(s)
+			if err != nil {
+				return err
 			}
+			nodeID, err := curNodeIdentifier(s, kubeDeps)
+			if err != nil {
+				return err
+			}
+			ncc = nodeconfig.NewNodeConfigController(client, nodeID, nodeConfigDir, flagCfg)
+		} else {
+			// pass a nil client to disable fetching configuration dynamically
+			ncc = nodeconfig.NewNodeConfigController(nil, "", nodeConfigDir, flagCfg)
 		}
-	}
 
-	// Validate configuration.
-	if err := validateConfig(s); err != nil {
-		return err
+		// when you run the controller, either you get back a valid config to use, or the Kubelet crashes because
+		// something was fatally wrong with the configuration. Non-fatal errors will be logged, but not returned from Run().
+		kcToUse := ncc.Run()
+
+		// Update s (KubeletServer) to any new config
+		kcToUseInternal := componentconfig.KubeletConfiguration{}
+		err := api.Scheme.Convert(kcToUse, &kcToUseInternal, nil)
+		if err != nil {
+			return err
+		}
+		s.KubeletConfiguration = kcToUseInternal
```

Comment:
A few thoughts on the structure of this:
* can we separate the dynamic config setup into a separate function/file to make the inputs (config and client) and outputs (controller and resulting config) clearer?
* mutating the config in the KubeletServer object makes this hard to reason about. can you hold the dynamically fetched config locally and propagate it to where it needs to go instead?
* I would expect all the dynamic config lookup to happen *after* the BootstrapKubeconfig handling... otherwise, bootstrapped kubelets won't have a client yet.